kart/spatial_tree.py

```diff
@@ -0,0 +1,430 @@
+    commit_spec - a list of commits to index (ancestors of these are implicitly included).
+        Commits can be exluded by prefixing with '^' (ancestors of these are implicitly excluded).
+        (See git rev-list for the full list of possibilities for specifying commits).
+    verbosity - how much non-essential information to output.
+    clear_existing - when true, deletes any pre-existing data before re-indexing.
+    """
+    import pywraps2 as s2
+
+    crs_helper = CrsHelper(repo)
+    feature_oid_iter = iter_feature_oids(repo, commit_spec)
+
+    s2_coverer = s2.S2RegionCoverer()
+    s2_coverer.set_max_cells(S2_MAX_CELLS_INDEX)
+    s2_coverer.set_max_level(S2_MAX_LEVEL)
+
+    progress_every = None
+    if verbosity >= 1:
+        progress_every = max(100, 100_000 // (10 ** (verbosity - 1)))
+
+    db_path = repo.gitdir_file(KartRepoFiles.S2_INDEX)
+    engine = sqlite_engine(db_path)
+    with sessionmaker(bind=engine)() as sess:
+        if clear_existing:
+            drop_tables(sess)
+
+        SpatialTreeTables.create_all(sess)
+
+    click.echo(f"Indexing {' '.join(commit_spec)} ...")
+    t0 = time.monotonic()
+    i = 0
+
+    # Using sqlite directly here instead of sqlalchemy is about 10x faster.
+    # Possibly due to huge number of unbatched queries.
+    # TODO - investigate further.
+    db = sqlite.connect(f"file:{db_path}", uri=True)
+    with db:
+        dbcur = db.cursor()
+
+        for i, (ds_path, feature_oid) in enumerate(feature_oid_iter):
+            if i and progress_every and i % progress_every == 0:
+                click.echo(f"  {i:,d} features... @{time.monotonic()-t0:.1f}s")
+
+            transforms = crs_helper.transforms_for_dataset(ds_path)
+            if not transforms:
+                continue
+            geom = get_geometry(repo, feature_oid)
+            if geom is None:
+                continue
+            try:
+                s2_cell_tokens = find_s2_cells(s2_coverer, geom, transforms)
+            except Exception as e:
+                L.warning(f"Couldn't locate S2 cells for {feature_oid}:\n{e}")
+                continue
+
+            params = (bytes.fromhex(feature_oid),)
+            row = dbcur.execute(
+                "SELECT rowid FROM blobs WHERE blob_id = ?;", params
+            ).fetchone()
+            if row:
+                rowid = row[0]
+            else:
+                dbcur.execute("INSERT INTO blobs (blob_id) VALUES (?);", params)
+                rowid = dbcur.lastrowid
+
+            if not s2_cell_tokens:
+                continue
+
+            params = [(rowid, token) for token in s2_cell_tokens]
+            dbcur.executemany(
+                "INSERT OR IGNORE INTO blob_cells (blob_rowid, cell_token) VALUES (?, ?);",
+                params,
+            )
+
+    t1 = time.monotonic()
+    click.echo(f"Indexed {i} features in {t1-t0:.1f}s")
+
+
+NO_GEOMETRY_COLUMN = object()
+
+
+def get_geometry(repo, feature_oid):
```

Comment:
Presumably we have this functionality somewhere else? Can we re-use it?

Reply:
This one works a bit different - mostly we load the repository at a particular revision. Then every feature we try to read has an attached legend ID, and we can combine that legend with the current schema to see which one is the geometry column. However, in this case, I'm running a magic git command that iterates through all features added in a certain set of commits. Its not straight-forward when using this command to see exactly which commit the feature was added at, so its not obvious what should be done in order to look up the attached legend or the current schema. But, it is possible to find the geometry field without doing those things, so that's what this code does.